src/graph/Graph.java

```diff
@@ -0,0 +1,172 @@
+package graph;
+
+import java.util.*;
+
+public class Graph {
+
+  private Map<Integer, Vertex> vertices;
+  private String name;
+
+  public Graph() {
+    this("unnamed");
+  }
+
+  public Graph(String name) {
+    this.vertices = new HashMap<>();
+    this.name = name;
+  }
+
+  public void setName(String name) {
+    this.name = name;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public Vertex getNode(int id) {
+    return this.vertices.get(id);
+  }
+
+  public Set<Integer> getNodesID() {
+    return vertices.keySet();
+  }
+
+  public Collection<Vertex> getNodes() {
+    return vertices.values();
+  }
+
+  public Vertex addNode(Vertex v) {
+    this.vertices.put(v.getId(), v);
+    return v;
+  }
+
+  public Vertex addNode(int id) {
+    Vertex node = new Vertex(id);
+    return this.addNode(node);
+  }
+
+  public Vertex addNode() {
+    if (this.isEmpty()) {
+      return this.addNode(0);
+    }
+    int max_id = Collections.max(this.vertices.keySet());
+    return this.addNode(max_id + 1);
+  }
+
+  public void removeNode(int id) {
+    for (Vertex v : this.getNodes()) {
+      v.removeNeighbor(id);
+    }
+    this.vertices.remove(id);
+  }
+
+  public void removeNode(Vertex v) {
+    this.removeNode(v.getId());
+  }
+
+  public Edge addEdge(Vertex node_u, Vertex node_v) {
```

Comment:
Would recommend adding edges directly.

Reply:
Edge object

Reply:
Judging from the behaviour of adding edges, this should be an undirected graph. Let's start with a directed version, and the undirected graph can be easy inheritated from this Graph structure. Or if you want a Base Graph (can be a interface) and fork inheritances for directed graph and undirected graph. Up to you. But a good programming style is make it clear when you are coding.